### PR TITLE
Some optimisations to two_stage_method

### PR DESCRIPTION
**Pull request by Vaibhavdixit02**

### CODE REVIEW

File: src/two_stage_method.jl [hunk 1]
```diff
@@ -78,24 +93,33 @@ function two_stage_method(prob::DiffEqBase.DEProblem,tpoints,data;kernel= :Epane
...
     n = length(tpoints)
     h = (n^(-1/5))*(n^(-3/35))*((log(n))^(-1/16))
-    estimated_solution = zeros(n,size(data)[1])
-    estimated_derivative = zeros(n,size(data)[1])
+    estimated_solution = zeros(size(data)[1],n)
+    estimated_derivative = zeros(size(data)[1],n)
     kernel_function = decide_kernel(kernel)
     e1 = [1;0]
     e2 = [0;1;0]
+    global du = similar(prob.u0) # have to adjust type for autodifferentiation
```
Comment (ChrisRackauckas) [inline src/two_stage_method.jl]: don't use globals. Why would you mark this global anyways?

File: src/DiffEqParamEstim.jl [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module DiffEqParamEstim
 using DiffEqBase, LsqFit, PenaltyFunctions,
-      RecursiveArrayTools, ForwardDiff, Calculus, Distributions, BlackBoxOptim, LinearAlgebra
+      RecursiveArrayTools, ForwardDiff, Calculus, Distributions, BlackBoxOptim, LinearAlgebra, Optim
```
Comment (ChrisRackauckas) [inline src/DiffEqParamEstim.jl]: why is this necessary?

File: src/two_stage_method.jl [hunk 3]
```diff
@@ -98,20 +98,20 @@ function two_stage_method(prob::DiffEqBase.DEProblem,tpoints,data;kernel= :Epane
...
     e2 = [0;1;0]
-    global du = similar(prob.u0) # have to adjust type for autodifferentiation
-    global sol = Vector{typeof(du)}(undef,n)
+    du = similar(prob.u0) # have to adjust type for autodifferentiation
+    sol = Vector{typeof(du)}(undef,n)
     if mpg_autodiff
-      global du_cache = DiffCache(du)
-      global sol_cache = DiffCache(sol)
+      du_cache = DiffCache(du)
+      sol_cache = DiffCache(sol)
```
Comment (Vaibhavdixit02) [inline src/two_stage_method.jl]: @ChrisRackauckas this needs to be changed as `sol` is not compatible with the `DiffCache` definition.

Comment (ChrisRackauckas) [reply src/two_stage_method.jl]: yes

Comment (Vaibhavdixit02) [reply src/two_stage_method.jl]: Can you suggest how to resolve this? I have tried a lot of things but haven't been able to get this working.

File: src/build_loss_objective.jl [hunk 4]
```diff
@@ -5,6 +5,13 @@ struct DiffEqObjective{F,F2} <: Function
   cost_function2::F2
 end
 
+function diffeq_sen_full(f, u0, tspan, p, t; alg=Tsit5(), sensalg=SensitivityAlg(), kwargs...)
+  prob = ODELocalSensitivityProblem(f,u0,tspan,p)
```
Comment (ChrisRackauckas) [inline src/build_loss_objective.jl]: no, there isn't a use for the forward sensitivities given the test results

File: src/build_loss_objective.jl [hunk 5]
```diff
@@ -49,9 +56,14 @@ function build_loss_objective(prob::DiffEqBase.DEProblem,alg,loss,regularization
   if mpg_autodiff
     gcfg = ForwardDiff.GradientConfig(cost_function, autodiff_prototype, autodiff_chunk)
     g! = (x, out) -> ForwardDiff.gradient!(out, cost_function, x, gcfg)
+  elseif lsa_gradient && typeof(loss) == L2Loss
+    sol_,sens = diffeq_sen_full(prob.f,pro.u0,pro.tspan,loss.t)
```
Comment (ChrisRackauckas) [inline src/build_loss_objective.jl]: you can't closure the problem because it changes.

File: src/two_stage_method.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/two_stage_method.jl]: sol should be an array of DiffCache

File: src/two_stage_method.jl [hunk 6]
```diff
@@ -78,24 +93,33 @@ function two_stage_method(prob::DiffEqBase.DEProblem,tpoints,data;kernel= :Epane
...
+        end 
         f = prob.f
         for i in 1:n
-          f(du,estimated_solution[i,:],p,tpoints[i])
+          est_sol = @view estimated_solution[:,i]
+          f(du,est_sol,p,tpoints[i])
           sol[i] = copy(du)
         end
-        sqrt(sum((vec(estimated_derivative') .- vec(VectorOfArray(sol))).^2))
+        sqrt(sum((vec(estimated_derivative) .- vec(VectorOfArray(sol))).^2))
```
Comment (ChrisRackauckas) [inline src/two_stage_method.jl]: sum(abs2(...)) and use a generator

File: src/two_stage_method.jl [hunk 7]
```diff
@@ -78,24 +93,33 @@ function two_stage_method(prob::DiffEqBase.DEProblem,tpoints,data;kernel= :Epane
...
-
     # Step - 2
     cost_function = function (p)
-        # have to adjust type for autodifferentiation
-        du = similar(prob.u0, promote_type(eltype(prob.u0), eltype(p)))
-        sol = Vector{typeof(du)}(undef,n)
+        if mpg_autodiff
+          du_ = get_tmp(du_cache,eltype(p))
+          du = reinterpret(eltype(p),du_)
+          sol_ = map(i -> get_tmp(i,eltype(p)),sol_cache)
```
Comment (Vaibhavdixit02) [inline src/two_stage_method.jl]: @ChrisRackauckas  here I am getting `LoadError: UndefRefError: access to undefined reference` which I am not sure ow to avoid, is it because of using `undef` in the `Vector` initialisation?

Comment (ChrisRackauckas) [reply src/two_stage_method.jl]: yes.

**Comment by Vaibhavdixit02** [comment]
@ChrisRackauckas please take a look at this
